Dockerfile

```diff
@@ -73,6 +72,8 @@ ADD web /ZoneMinder/web/
 ADD cmakecacheimport.sh CMakeLists.txt version zm.conf.in zmconfgen.pl.in zmlinkcontent.sh.in zoneminder-config.cmake /ZoneMinder/
 ADD conf.d /ZoneMinder/conf.d
 
+RUN git clone --depth 1 -b 3.0 https://github.com/FriendsOfCake/crud.git /ZoneMinder/web/api/app/Plugin/Crud
```

Comment:
Rather than git clone crud separately, do this instead:
```git submodule update --init --recursive```

This ensures you end up with the version of crud the project is intended to be used with (currently 3.0.10) and will pull down any other submodules we might add in the future.

Reply:
This doesn't look like an option when you are building the image.  Always results in an error.  I know this is noted as a requirement if you build the package but in an automated build for docker hub, it's not an option.

An alternative is to make travis ci build and push out to docker hub.  

It's up to you guys, if you leave it, it can be an automated build on docker hub.  Or we can make travis ci push it out.

Reply:
I'd rather not hardcode this. If `git clone` works here then `git submodule update` should work just as well. Try calling this command a couple lines down, after WORKDIR is set to ZoneMinder. Alternatively, you can specify the path directly in the command: `git submodule update --init --recursive /ZoneMinder`. If you can error, then let us know what it is.

Reply:
This is what I see when I attempt to add the .git directory and the .gitmodules file:
```
Step 14/27 : ADD .git /ZoneMinder/.git
 ---> 06223b78c962
Step 15/27 : ADD .gitmodules /ZoneMinder/.gitmodules
 ---> 184287c58e62
Step 16/27 : RUN git submodule update --init --recursive /ZoneMinder.
 ---> Running in e6073273071e
fatal: Not a git repository (or any of the parent directories): .git
The command '/bin/sh -c git submodule update --init --recursive /ZoneMinder.' returned a non-zero code: 128
```

Hopefully just a simple thing I'm doing wrong.

Reply:
Sorry for the delay. I did some looking into and it seems you can git checkout the project and get the submodules all with one command:
`git clone --recursive https://github.com/ZoneMinder/ZoneMinder`

Seems to work for me. I'll test it in docker...

Reply:
Nice, testing now.  Looks to work well.  

I'm going to test something else real quick due to having issues with high resolutions, I think it needs [this](https://forums.zoneminder.com/viewtopic.php?t=25855).  I'll play a little bit and get at least this needed change in there.

Reply:
Yeah, alright. It's building on my end. 

I ended up removing the block of ADD's, which copied the local code into the container. Git clone does that for us now.

Further down, you may want to change:
`ADD utils/docker/start.sh /tmp/start.sh`
to
`RUN cp -v utils/docker/start.sh /tmp/start.sh`

I think that is the right syntax. Not sure why we are putting start.sh into /tmp. Maybe that's for speed since that folder is likely in ram.

Then near the end, we no longer need this:
`ADD utils/docker /ZoneMinder/utils/docker/`

This folder will be there already thanks to the git clone.

This of course makes the whole container portable... you now only need the Dockerfile and not the rest of the repo.

Reply:
Here is a gist showing the whole thing after changes:
{% gist 6a83fbf30c82235bdc63de59214a8ddf %}